src/twisted/internet/defer.py

```diff
@@ -1266,9 +1281,9 @@ def __init__(
         """
         self._deferredList = list(deferredList)
 
-        self.resultList: List[Optional[Tuple[bool, _DeferredResultT]]] = [None] * len(
-            self._deferredList
-        )
+        self.resultList: _DeferredListResultListT = [
+            cast(_DeferredListResultItemT, None)
```

Comment:
Casting to `None` here instead of typing `self.resultList` as optional because it's only allowed to be set to `None` here.

Reply:
I hadn't realized that the client might be looking at the `resultList` property rather than using the callback argument. I think the latter would be more common and more in line with Twisted's design, but both are valid.

Reply:
Maybe add a comment explaining this decision in the code.

Reply:
Yeah, I agree it's more common.  Because `resultList` is a public attribute, I think it's more important to type it in a way that makes sense to the client over the implementation.

(I have a feeling it might be better if it weren't public, but I'm not a big `DeferredList` user so I don't think I know enough about that.)

Reply:
The `DeferredList` class should probably be replaced by two functions, one for each value of `fireOnOneCallback`. But that would have to go through a lengthy deprecation cycle, so even if someone would write those replacement functions today, having `DeferredList` annotated is going to be useful for the coming years.

Reply:
Agreed.

Reply:
OK I'm looking at this again, and the `None` values to show to the caller when the child `Deferred` objects have yet to fire, so I'm switching this back to optional, which requires a bit more checking elsewhere.  See commit 02110624550737ef58c34f623b5dd36d02f04479.

Reply:
@mthuurne Can you sanity check that commit?

Reply:
It looks mostly fine, except that the type of the value returned by `DeferredList` should not contain `Optional`, since when the callback fires, all the `None` elements have been replaced. In other words, `resultList` contains a partial result and can therefore contain `None`, but the list returned via the callback is a final result and cannot contain `None`.

Reply:
OK done in d0d586ae287c6feaacb0c094a3bc9b1673dc890c